torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -25,58 +26,135 @@ static int ceilDiv(const int a, const int b) {
   return (a + b - 1) / b;
 }
 
+// IO data structure for kernel code;
+static auto typeinfo = R"(
+template<typename T, int N>
+struct Tensor {
+  T& operator[](int ind) {
+    return data[ind];
+  };
+
+  T* data;
+  int size[N];
+  int stride[N];
+};
+)";
+
+// include IO data structure for host code
+struct KernelArgumentHolder {
+  std::vector<void*> arguments;
+  std::vector<char> buffer;
+  void* buffer_ptr;
+
+  KernelArgumentHolder(size_t n_inputs, size_t n_dimension_per_tensor) {
+    arguments.reserve(n_inputs);
+    // We are being generous here on the allocated buffer;
+    buffer.resize(
+        n_inputs * (sizeof(void*) + 2 * sizeof(int) * n_dimension_per_tensor));
+    buffer_ptr = buffer.data();
+  }
+
+  void** args() {
+    return arguments.data();
+  }
+
+  // this should comply to the storage of Tensor object defined in typeinfo;
+  void push_tensor(
+      const at::Tensor& val,
+      c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
+    arguments.push_back(buffer_ptr);
+
+    // passing address, type doesn't really matter here;
+    auto data_ptr = static_cast<void**>(buffer_ptr);
+    *data_ptr = val.data_ptr();
+    buffer_ptr += sizeof(char*);
+
+    if (broadcasted_size) {
```

Comment:
Is this a supported way to use c10::optional? I shouldn't be surprised, but I haven't seen this use before. I've always used `if(... == c10::nullopt)`

Reply:
I'm generally a bit confused about the logic contained in here. I think it would be helpful to go through this together.

Reply:
It's used a lot in core. https://github.com/pytorch/pytorch/blob/9ada7abc18fcc6d575d1c5e801e85666d634b412/aten/src/ATen/core/jit_type.h#L389-L402

If we want to play it safe, we can always do `broadcasted_size.has_value()`

Reply:
I think it's fine, I just wasn't aware. Thanks.